### ISSUE TITLE
Wrapping generic ClassFileTransformer into ResettableClassFileTransformer

### ISSUE THREAD

**Issue by kriegaex** [raphw/byte-buddy#870]
### Preface

I like ByteBuddy (BB) for many reasons, one of which is that I can dynamically reset transformations on demand when using a `ResettableClassFileTransformer`. As the transformation I am going to talk about is not changing the class structure (i.e. retransformation back and forth should be permitted), we have no problem there.

### Situation

I am doing some kind of constructor code transformation which is not possible directly in BB and which I also failed to implement via ASM class visitor. That is too bad because it means I cannot just use an ASM class visitor wrapper. I know that what I want is possible with ASM in theory, but my Java byte code knowledge is not good enough and I failed miserably. I also need to generate different types of code for different types of constructors and it would take me weeks to get to a point of success. The last two days with ASM were a nightmare.

Now today I started playing with Javassist and was able to implement exactly what I wanted within a few hours, mostly because Javassist permits mehr to pass Java code as a string and it takes care of compiling it. Also its API with regard to identifying constructors in classes is easy to handle for me. Probably the latter would also be easy with the BB tree API, but I only used the visitor API so far. Anyway, the byte code generation part is the tricky one, as I said.

So now the tool I am building is kind of a hybrid of

  * ByteBuddy (using advice API),
  * ASM base classes (using class visitor API),
  * Javassist (constructor instrumentation).

Until the day I might feel like migrating the Javassist stuff to ASM, I would like to wrap it into a `ResettableClassFileTransformer` and just delegate the actual transformation to Javassist, similar to how it is possible for ASM visitors. BTW, if it is relevant, my Javassist transformer implements `java.lang.instrument.ClassFileTransformer` already, i.e. I can use it from a Java agent or attach the transformer dynamically via Java Attach API, e.g. via `ByteBuddyAgent.install().addTransformer(..)`.

### Question

Can you please tell me how I would go about making my Javassist (or any other non-BB type of) transformation resettable?

**Comment by raphw** [comment]
I would not know how to integrate the two. Javassist has to serialize the entire type tree to work. The reason you can use ASM is due to Byte Buddy's dependance on it where the underlying API gets exposed.

Can you show me what Javassist code you are implementing? This should very much be possible with ASM.

**Comment by kriegaex** [comment]
> This should very much be possible with ASM.

Yes, of course. But as I said:

>> I know that what I want is possible with ASM in theory, but my Java byte code knowledge is not good enough and I failed miserably.

> I would not know how to integrate the two.

I can do it manually by unregistering the corresponding transformer and doing a retransformation of all affected classes. I just wanted to avoid the manual bookkeeping involved. So I guess I have to come up with a solution of my own.

> The reason you can use ASM is due to Byte Buddy's dependance on it

I know. In my "remove final" module I started with plain ASM and then migrated to using ASM embedded in BB in order to enjoy the benefits of its infrastructure, such as resettable transformers and BB Maven plugin. 😉

> Can you show me what Javassist code you are implementing? 

I can, indeed. Maybe you have read my messages on Gitter and the corresponding [StackOverflow link](https://stackoverflow.com/q/61967586/1082681) I sent you there. Basically it is the idea I described there, but I implemented in a slightly different way: For non-mockable classes (final) or for situation where I cannot inject a mock and need to create a "global" mock by stubbing out the constructor I want to inject code into constructors of my target classes (and all their super classes) which avoids all side effects by bypassing the original constructor code and calling super constructors with dummy values (`null`, `0`, `false`) instead, resulting in uninitialised objects the methods of which you can then additionally sub to return whatever results you like using another part of my framework (utilising BB advice, by the way). This also works for already loaded (bootstrap or application) classes because it only changes code inside of constructors and methods. This is how I do the constructor part (because I cannot use BB advices in order to skip constructors):

```java
  @Override
  public byte[] transform(ClassLoader loader, String className, Class<?> classBeingRedefined, ProtectionDomain protectionDomain, byte[] classfileBuffer) {
    CtClass ctClass;
    try {
      // Caveat: Do not just use 'classPool.get(className.replaceAll("/", "."))' because we would miss
      // previous transformations. It is necessary to really parse 'classfileBuffer'.
      ctClass = classPool.makeClass(new ByteArrayInputStream(classfileBuffer));
    }
    catch (IOException e) {
      return null;
    }
    if (!shouldTransform(ctClass))
      return null;
    applyTransformations(ctClass);
    try {
      return ctClass.toBytecode();
    }
    catch (Exception e) {
      System.out.println(e);
      return null;
    }
  }

  private void makeGloballyMockable(CtClass ctClass) throws NotFoundException, CannotCompileException {
    if (LOG_GLOBAL_MOCK)
      log("Adding global mock capability to class " + ctClass.getName());
    String superCall = getSuperCall(ctClass);
    for (CtConstructor ctConstructor : ctClass.getDeclaredConstructors()) {
      if (LOG_GLOBAL_MOCK)
        log("Adding global mock capability to constructor " + ctConstructor.getLongName());
      ctConstructor.insertBefore(
        String.join("\n",
          "if (de.scrum_master.agent.global_mock.GlobalMockRegistry.isObjectInConstructionMock()) {",
          "  " + superCall,
          "  return;",
          "}"
        )
      );
      // Alternative idea by Björn Kautler without stack trace generation + parsing:
/*
      if (GlobalMockRegistry.isCurrentlyInitializingMockThreadLocal()) {
        super(null, 0, false);
        return;
      }
      if (GlobalMockRegistry.isMock(Super.class)) {
        GlobalMockRegistry.setCurrentlyInitializingMockThreadLocal(true);
        try {
          super(null, 0, false);
          return;
        }
        finally {
          GlobalMockRegistry.setCurrentlyInitializingMockThreadLocal(false);
        }
      }
*/
    }
  }

  private String getSuperCall(CtClass childClass) throws NotFoundException {
    CtClass superClass = childClass.getSuperclass();
    // Get declared non-private constructors (we cannot call private ones via 'super()')
    CtConstructor[] superConstructors = superClass.getConstructors();
    assert superConstructors.length > 0
      : "There has to be at least one accessible super class constructor";
    // Use first super constructor found
    // TODO(?): use constructor with shortest parameter list (but we have to solve the general case anyway)
    CtConstructor superConstructor = superConstructors[0];
    String parameterList = Stream.of(superConstructor.getParameterTypes())
      .map(parameterType -> {
        switch (parameterType.getName()) {
          case "byte":
            return "0";
          case "short":
            return "0";
          case "int":
            return "0";
          case "long":
            return "0L";
          case "float":
            return "0.0f";
          case "double":
            return "0.0d";
          case "boolean":
            return "false";
          default:
            return "null";
        }
      })
      .collect(Collectors.joining(", "));
    return "super(" + parameterList + ");";
  }
```

So you see, I have a simple `if` and subsequequent `super()` call which is super class specific and dynamic in nature, thus the generator method. Then I just call `return`. The more efficient scheme proposed by Björn in our private chat would also be more complex in Java terms, but allow me to work without parsing call stacks and also make it easier to work with the same classes on different classloaders which once are to be mocked and once not. Anyway, speed or the more complex application cases are not my main concern at the moment, as long as my first proof of concept just works. And it does, beautifully.

In either of the two implementation variants, the Java code looks pretty simple, but in ASM this would be a complex list of opcodes which is beyond my abilities for now. I am sure you could quite easily do it because you are an experienced ASM user.

Besides, this is the code called by the code woven into the constructor, if you are interested:

```java
  public static boolean isObjectInConstructionMock() {
    // TODO: This is slow and assumes that global mocks should be mocked on all class loaders.
    //       Consider Björn Kautler's ThreadLocal idea or an alternative approach?
    // This is said to be faster than Thread.currentThread().getStackTrace()
    StackTraceElement[] stackTrace = new Throwable().getStackTrace();
    assert stackTrace.length > 1;
    for (int i = 1; i < stackTrace.length; i++) {
      StackTraceElement stackTraceElement = stackTrace[i];
      if (!stackTraceElement.getMethodName().equals("<init>"))
        break;
      if (isMock(stackTraceElement.getClassName()))
        return true;
    }
    return false;
  }
```

**Comment by raphw** [comment]
Sorry for not following this up, it slipped me. Did you resolve this issue in the mean time by what we discussed in the chat?

**Comment by kriegaex** [comment]
Well, I did not wrap the Javassist transformer into a BB one, but took care of making the transformation resettable by myself, adding some internal state to the transformer.

As for migrating from Javassist to BB, I managed to do it as I explained in the chat: using Javassist-generated byte code as an input for ASMifier, then using the output of that as an input for an ASM Class Visitor, then using a BB ASM Class Visitor Wrapper around that. Later I also managed to get rid of the Class Visitor + Wrapper altogether in favour of a simple Method Visitor + Wrapper. Now it is working just as well as in Javassist, but I have one less dependency in my tool mix.

Thanks for all your feedback, encouragement and inspiration.
